### DISCUSSION TITLE
Parametrize global variables best practice

### DISCUSSION THREAD

**Discussion by TerranWorks** [pytest-dev/pytest#10182] [Q&A]
Hello, I'm using pytest to test my sqlalchemy backend and I found that for most of my tests, i need to see how my code behaves when a speficied user exists and when it doesn't. 
I keep having to write this check manually at the beginning of each test but it occured to me that if I could create two users at the beginning of the testing session, only actually create one of them, and then pass both to my tests through the parametrize decorator then i could conveniently check both cases. 
But here are the problems:
- I need the users to be created at the beginning and cleaned up at the end which I would like to do with a fixture with session scope but I couldn't get to work
- I need the users to be created only after I've tested the user creation method. I was using the depends plugin for this and I think that part was working

[This](https://miguendes.me/how-to-use-fixtures-as-arguments-in-pytestmarkparametrize) was so close to being what I wanted but I kept getting ``TypeError: object of type 'LazyFixture' has no len()``. My fixture returns a tuple by the way.

I know this is a little obscure but I have to imagine I'm not the first person to want to do something like this. Does anyone have any suggestions?

**Comment by RonnyPfannschmidt** [comment]
Without an more detailed example we can't give any suggestion

Please show a example piece of code that demonstrates what you want to work so we can suggest a fit solution/workaround

**Comment by nicoddemus** [comment]
An example like @RonnyPfannschmidt asked would make things more clear, but from what I can gather, I think you can use a [parametrized fixture](https://docs.pytest.org/en/stable/how-to/fixtures.html#fixture-parametrize) to solve your problem.

**Comment by TerranWorks** [comment]
So here is what I have so far and what I'm trying to achieve.

I have a two fixtures with session scope. I generate two sets of user data (a tuple of strings: username, email) but only actually create one of the users. Notice that setup_user depends on ``test_create_account``, since that function should probably be tested before it is used within the fixture
```
@pytest.mark.depends(on="test_create_account")
@pytest.fixture(scope="session")
def setup_user(connection):
    user=generate_unique_user()
    username,email=user
    proc.create_account(connection, username, email)
    yield user

    #drop user after session 

@pytest.fixture(scope="session")
def setup_fake_user():
    fake_user=generate_unique_user()
    yield fake_user
```

I then want some way of passing both ``user`` and ``fake_user``  as parameters to some of my tests so they can be run once with a user that exists and again with a user that doesn't

Criteria:
- Any users that are created should be destroyed by the end. I'd like to create the user at the beginning and keep reusing it but if i have to make a new one for each test that is okay
- Any test or fixture that uses the ``create_user`` method should depend on ``test_create_user``
- user and fake_user should be able to be used as parameters for my tests so I can run it once with an existing user and again with a non-existent user which I expect to break everything

Its really a shame that [this](https://miguendes.me/how-to-use-fixtures-as-arguments-in-pytestmarkparametrize) didnt work because its so close to what I want
